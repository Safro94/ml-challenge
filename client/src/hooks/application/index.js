import React, {
	createContext,
	useContext,
	useReducer,
	useCallback,
} from 'react';

const SET_SEARCH_TERM = 'SET_SEARCH_TERM';
const SET_RESULT = 'SET_RESULT';

const ApplicationContext = createContext();

const useApplication = () => {
	const [application, setApplication] = useContext(ApplicationContext);

	const setSearchTerm = useCallback(
		term => setApplication({ type: SET_SEARCH_TERM, value: term }),
		[setApplication]
	);

	const setResult = useCallback(
		result => setApplication({ type: SET_RESULT, value: result }),
		[setApplication]
	);

	return { ...application, setSearchTerm, setResult };
};

const ApplicationProvider = ({ application = {}, children }) => {
	const reducer = (state, action) => {
		switch (action.type) {
			case SET_SEARCH_TERM:
				return { ...state, term: action.value };
			case SET_RESULT:
				return {
					...state,
					items: action.value.items,
					categories: action.value.categories,
				};
			default:
				return state;
		}
	};

	return (
		<ApplicationContext.Provider value={useReducer(reducer, application)}>
			{children}
		</ApplicationContext.Provider>
	);
};

export { useApplication, ApplicationProvider };
