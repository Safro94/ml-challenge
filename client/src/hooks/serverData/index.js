import { useState, useEffect } from 'react';

/*
  Custom hook creado para manejar la data que se fetchea
  server side. Se crea hook para que cada componente que
  requiera data del server pueda usarlo y no tener que
  tener esto repetido en cada componente
*/
export const useServerSideData = staticContext => {
	const [data, setData] = useState(
		staticContext ? staticContext.initialData : window.__initialData__
	);

	useEffect(() => {
		const initState = document.getElementById('initState');
		delete window.__initialData__;

		if (initState && initState.parentNode) {
			initState.parentNode.removeChild(initState);
		}
	}, []);

	return data;
};
