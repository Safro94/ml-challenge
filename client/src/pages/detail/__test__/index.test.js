import React from 'react';

import { render, screen } from '@testing-library/react';

import DetailContainer from 'containers/detail';
import BreadcrumbContainer from 'containers/breadcrumb';

import { useServerSideData } from 'hooks/serverData';

import Detail from '../';

const mockId = 10;
jest.mock('react-router-dom', () => ({
	useParams: () => ({ id: mockId }),
}));

const mockItem = { id: 10, title: 'title' };
const mockCategories = ['category', 'category'];
jest.mock('hooks/serverData', () => ({
	useServerSideData: jest.fn(),
}));

jest.mock('utils/currency', () => jest.fn(() => '300,50'));

jest.mock('containers/detail', () => {
	return jest.fn(({ data }) => <div>{data?.item?.title}</div>);
});

jest.mock('containers/breadcrumb', () => {
	return jest.fn(data => (
		<div>
			{data?.categories?.map((category, index) => (
				<span key={index}>{category}</span>
			))}
		</div>
	));
});

describe('Detail', () => {
	it('should render the items if staticContext.initialData has data', () => {
		//Arrange
		const staticContext = {
			initialData: {
				item: mockItem,
				categories: mockCategories,
			},
		};

		useServerSideData.mockImplementation(() => ({
			categories: mockCategories,
			item: mockItem,
		}));

		//Act
		render(<Detail staticContext={staticContext} />);

		//expect
		expect(screen.getByText(mockItem.title)).toBeInTheDocument();
		expect(screen.queryAllByText('category')).toHaveLength(2);

		expect(DetailContainer).toHaveBeenCalledTimes(1);
		expect(BreadcrumbContainer).toHaveBeenCalledTimes(1);

		expect(DetailContainer.mock.calls[0][0].data).toEqual(
			staticContext.initialData
		);
		expect(DetailContainer.mock.calls[0][0].itemId).toBe(mockId);

		expect(BreadcrumbContainer.mock.calls[0][0].categories).toEqual(
			staticContext.initialData.categories
		);
	});

	it('should not render the items if staticContext.initialData has no data', () => {
		//Arrange
		useServerSideData.mockImplementation(() => null);

		//Act
		render(<Detail />);

		//expect
		expect(screen.queryByText(mockItem.title)).not.toBeInTheDocument();
		expect(screen.queryAllByText('category')).toHaveLength(0);

		expect(DetailContainer).toHaveBeenCalledTimes(1);
		expect(BreadcrumbContainer).toHaveBeenCalledTimes(1);

		expect(DetailContainer.mock.calls[0][0].data).toBeNull();
		expect(DetailContainer.mock.calls[0][0].itemId).toBe(mockId);

		expect(BreadcrumbContainer.mock.calls[0][0].categories).not.toBeDefined();
	});
});
