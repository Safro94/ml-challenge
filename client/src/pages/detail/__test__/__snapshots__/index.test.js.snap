// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Detail should render OK 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="container"
    >
      <ul />
    </div>
    <div
      class="container"
      data-testid="loading"
      style="height: 200px; width: 100%;"
    >
      <div
        class="loader"
      />
    </div>
  </div>
</div>
`;
